cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasAnalysis.qll

```diff
@@ -96,60 +98,60 @@ IntValue getPointerBitOffset(PointerOffsetInstruction instr) {
-    exists(ConvertToDerivedInstruction convert |
-      convert = instr and
-      bitOffset = Ints::neg(Ints::mul(convert.getDerivation().getByteOffset(), 8))
-    ) or
-    // Converting to a virtual base class adds an unknown offset.
+predicate operandIsPropagated(Operand operand, IntValue bitOffset) {
+  exists(Instruction instr |
+    instr = operand.getInstruction() and
     (
-      instr instanceof ConvertToVirtualBaseInstruction and
-      bitOffset = Ints::unknown()
-    ) or
-    // Conversion to another pointer type propagates the source address.
-    exists(ConvertInstruction convert, Type resultType |
-      convert = instr and
-      resultType = convert.getResultType() and
+      // Converting to a non-virtual base class adds the offset of the base class.
+      exists(ConvertToBaseInstruction convert |
+        convert = instr and
+        bitOffset = Ints::mul(convert.getDerivation().getByteOffset(), 8)
+      ) or
+      // Converting to a derived class subtracts the offset of the base class.
+      exists(ConvertToDerivedInstruction convert |
+        convert = instr and
+        bitOffset = Ints::neg(Ints::mul(convert.getDerivation().getByteOffset(), 8))
+      ) or
+      // Converting to a virtual base class adds an unknown offset.
       (
-        resultType instanceof PointerType or
-        resultType instanceof Class  //REVIEW: Remove when all glvalues are pointers
-      ) and
-      bitOffset = 0
-    ) or
-    // Adding an integer to or subtracting an integer from a pointer propagates
-    // the address with an offset.
-    bitOffset = getPointerBitOffset(instr.(PointerOffsetInstruction)) or
-    // Computing a field address from a pointer propagates the address plus the
-    // offset of the field.
-    bitOffset = getFieldBitOffset(instr.(FieldAddressInstruction).getField()) or
-    // A copy propagates the source value.
-    tag instanceof CopySourceOperand and bitOffset = 0
+        instr instanceof ConvertToVirtualBaseInstruction and
+        bitOffset = Ints::unknown()
+      ) or
+      // Conversion to another pointer type propagates the source address.
+      exists(ConvertInstruction convert, Type resultType |
+        convert = instr and
+        resultType = convert.getResultType() and
+        (
+          resultType instanceof PointerType or
+          resultType instanceof Class  //REVIEW: Remove when all glvalues are pointers
+        ) and
+        bitOffset = 0
+      ) or
+      // Adding an integer to or subtracting an integer from a pointer propagates
+      // the address with an offset.
+      bitOffset = getPointerBitOffset(instr.(PointerOffsetInstruction)) or
+      // Computing a field address from a pointer propagates the address plus the
+      // offset of the field.
+      bitOffset = getFieldBitOffset(instr.(FieldAddressInstruction).getField()) or
+      // A copy propagates the source value.
+      operand instanceof CopySourceOperand and bitOffset = 0
+    )
```

Comment:
This might just be because of diff ugliness, but I'm having trouble convincing myself that all these cases are still covered

Reply:
Nothing significant has changed here. It's a bit easier to see if you use side-by-side diffs.